pkg/resources/quota/quota.go

```diff
@@ -0,0 +1,348 @@
+package quota
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	threescalev1 "github.com/3scale/3scale-operator/pkg/apis/apps/v1alpha1"
+	"github.com/integr8ly/integreatly-operator/apis/v1alpha1"
+	marin3rconfig "github.com/integr8ly/integreatly-operator/pkg/products/marin3r/config"
+	keycloak "github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1"
+	appsv1 "github.com/openshift/api/apps/v1"
+	appsv12 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ConfigMapData         = "quota-configs"
+	ConfigMapName         = "quota-config-managed-api-service"
+	RateLimitName         = "ratelimit"
+	BackendListenerName   = "backend_listener"
+	BackendWorkerName     = "backend_worker"
+	ApicastProductionName = "apicast_production"
+	ApicastStagingName    = "apicast_staging"
+	KeycloakName          = "rhssouser"
+	GrafanaName           = "grafana"
+)
+
+var (
+	// map of products iterate over that to build the return map
+	products = map[v1alpha1.ProductName][]string{
+		v1alpha1.Product3Scale: {
+			BackendListenerName,
+			BackendWorkerName,
+			ApicastProductionName,
+			ApicastStagingName,
+		},
+		v1alpha1.ProductRHSSOUser: {
+			KeycloakName,
+		},
+		v1alpha1.ProductMarin3r: {
+			RateLimitName,
+		},
+		v1alpha1.ProductGrafana: {
+			GrafanaName,
+		},
+	}
+)
+
+type Quota struct {
+	name            string
+	productConfigs  map[v1alpha1.ProductName]QuotaProductConfig
+	isUpdated       bool
+	rateLimitConfig marin3rconfig.RateLimitConfig
+}
+
+//go:generate moq -out product_config_moq.go . ProductConfig
+type ProductConfig interface {
+	Configure(obj metav1.Object) error
+	GetResourceConfig(ddcssName string) (corev1.ResourceRequirements, bool)
+	GetReplicas(ddcssName string) int32
+	GetRateLimitConfig() marin3rconfig.RateLimitConfig
+}
+
+var _ ProductConfig = QuotaProductConfig{}
+
+type QuotaProductConfig struct {
+	productName     v1alpha1.ProductName
+	resourceConfigs map[string]ResourceConfig
+	quota           *Quota
+}
+
+type ResourceConfig struct {
+	Replicas  int32                       `json:"replicas,omitempty"`
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type quotaConfigReceiver struct {
+	Name      string                        `json:"name,omitempty"`
+	RateLimit marin3rconfig.RateLimitConfig `json:"rate-limiting,omitempty"`
+	Resources map[string]ResourceConfig     `json:"resources,omitempty"`
+}
+
+func GetQuota(QuotaId string, QuotaConfig *corev1.ConfigMap, retQuota *Quota, isUpdated bool) error {
```

Comment:
Why are we passing in `retQuota *Quota` instead of returning `Quota` the usual way?

Reply:
we decided to follow a similar syntax as with the k8s client `Get` functions where a pointer is passed and only an error is returned.